CIRCexplorer2
-------------

[![Build](https://github.com/Yanglab/CIRCexplorer2/actions/workflows/python-package-conda.yml/badge.svg?branch=master)](https://github.com/YangLab/CIRCexplorer2/actions?query=workflow%3ABuild)
.. image:: https://github.com/Yanglab/CIRCexplorer2/actions/workflows/python-package-conda.yml/badge.svg?branch=master
    :target: https://github.com/YangLab/CIRCexplorer2/actions?query=workflow%3ABuild
.. image:: https://coveralls.io/repos/github/YangLab/CIRCexplorer2/badge.svg?branch=master
    :target: https://coveralls.io/github/YangLab/CIRCexplorer2?branch=master
.. image:: https://readthedocs.org/projects/circexplorer2/badge/?version=latest
    :target: http://circexplorer2.readthedocs.org/en/latest/?badge=latest
.. image:: https://img.shields.io/badge/license-MIT-orange.svg?style=flat-square
    :target: https://github.com/YangLab/CIRCexplorer2/blob/master/LICENSE.txt

CIRCexplorer2 is the successor of `CIRCexplorer`_ with plenty of new features to facilitate circular RNA identification and characterization.

.. _CIRCexplorer: http://yanglab.github.io/CIRCexplorer/

Features
--------

* Precisely annotate circular RNAs with high accuracy, good sensitivity and low memory consumption
* Support multiple circular RNA aligners (TopHat2/TopHat-Fusion, STAR, MapSplice, BWA and segemehl)
* *De novo* assemble novel circular RNA transcripts
* Characterize various of alternative splicing events of circular RNAs
* Fast identify circuar RNAs with STAR or BWA
* Support both single-read and paired-end sequencing.

CIRCexplorer2 documentation is available through https://readthedocs.org/ from `here`_, including installation instructions and tutorial.

.. _here: http://CIRCexplorer2.readthedocs.org

Citation
--------

Zhang XO\*, Dong R\*, Zhang Y\*, Zhang JL, Luo Z, Zhang J, Chen LL#, Yang L#. Diverse alternative back-splicing and alternative splicing landscape of circular RNAs. *Genome Res*, 2016, 26:1277-1287
